### PR TITLE
Async Snapshot Repository Deletes

### PR DESCRIPTION
**Pull request by original-brownbear**
Motivated by slow snapshot deletes reported in e.g. #39656 and the fact that these likely are a contributing factor to repositories accumulating stale files over time when deletes fail to finish in time and are interrupted before they can complete. 

* Makes snapshot deletion async and parallelizes some steps of the delete process that can be safely run concurrently via the snapshot thread poll
   * I did not take the biggest potential speedup step here and parallelize the shard file deletion because that's probably better handled by moving to bulk deletes where possible (and can still be parallelized via the snapshot pool where it isn't). Also, I wanted to keep the size of the PR manageable. 
* See https://github.com/elastic/elasticsearch/pull/39656#issuecomment-470492106
* Also, as a side effect this gives the `SnapshotResiliencyTests` a little more coverage for master failover scenarios (since parallel access to a blob store repository during deletes is now possible since a delete isn't a single task anymore).
* By adding a `ThreadPool` reference to the repository this also lays the groundwork to parallelizing shard snapshot uploads to improve the situation reported in #39657

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
+                writeIndexGen(updatedRepositoryData, repositoryStateId);
+            } catch (Exception ex) {
+                listener.onFailure(new RepositoryException(metadata.name(), "failed to delete snapshot [" + snapshotId + "]", ex));
+                return;
+            }
+            deleteSnapshotBlobs(snapshot, snapshotId, repositoryData, updatedRepositoryData, listener);
+        }
+    }
 
+    private void deleteSnapshotBlobs(SnapshotInfo snapshot, SnapshotId snapshotId, RepositoryData repositoryData,
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: `@Nullable SnapshotInfo snapshot`

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 2]
```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
+                                     RepositoryData updatedRepositoryData, ActionListener<Void> listener) {
+        final Executor executor = threadPool.executor(ThreadPool.Names.SNAPSHOT);
+        final AtomicInteger outstanding = new AtomicInteger(2);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Use `GroupedActionListener` instead?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 3]
```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
+        final ActionListener<Void> deleteListener = ActionListener.wrap(() -> {
+            if (outstanding.decrementAndGet() == 0) {
+                deleteIndices(snapshot, repositoryData, snapshotId, updatedRepositoryData, listener);
+            }
+        });
+        executor.execute(() -> {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: use ActionRunnable, which subclasses AbstractRunnable as `deleteSnapshotBlobIgnoringErrors` will only ignore IOException, not all other exceptions. We should make sure to always calls the listener.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 4]
```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
             // delete the snapshot file
             deleteSnapshotBlobIgnoringErrors(snapshot, snapshotId.getUUID());
+            deleteListener.onResponse(null);
+        });
+        executor.execute(() -> {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: use ActionRunnable

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 5]
```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
+            final RepositoryData updatedRepositoryData;
+            try {
+                repositoryData = getRepositoryData();
+                updatedRepositoryData = repositoryData.removeSnapshot(snapshotId);
+                writeIndexGen(updatedRepositoryData, repositoryStateId);
+            } catch (Exception ex) {
+                listener.onFailure(new RepositoryException(metadata.name(), "failed to delete snapshot [" + snapshotId + "]", ex));
+                return;
+            }
+            deleteSnapshotBlobs(snapshot, snapshotId, repositoryData, updatedRepositoryData, listener);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: can we do the chaining of the individual substeps here? I think it makes the flow clearer instead of jumping from one method to the next

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 6]
```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
             deleteGlobalMetaDataBlobIgnoringErrors(snapshot, snapshotId.getUUID());
+            deleteListener.onResponse(null);
+        });
+    }
 
-            // Now delete all indices
-            if (snapshot != null) {
-                final List<String> indices = snapshot.indices();
-                for (String index : indices) {
+    private void deleteIndices(SnapshotInfo snapshot, RepositoryData repositoryData, SnapshotId snapshotId,
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: `@Nullable SnapshotInfo snapshot`

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 7]
```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
+                               RepositoryData updatedRepositoryData, ActionListener<Void> listener) {
+        // Now delete all indices
+        if (snapshot != null) {
+            final List<String> indices = snapshot.indices();
+            if (indices.isEmpty()) {
+                deleteUnreferencedIndices(repositoryData, updatedRepositoryData, listener);
+                return;
+            }
+            final AtomicInteger outstandingIndices = new AtomicInteger(indices.size());
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: GroupedActionListener

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 8]
```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
+            for (String index : indices) {
+                threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(() -> {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: ActionRunnable

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: We can further parallelize this to the shard level I think (can be a follow-up)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 9]
```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
+            threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(() -> {
                 try {
                     indicesBlobContainer.deleteBlob(indexId.getId());
                 } catch (DirectoryNotEmptyException dnee) {
                     // if the directory isn't empty for some reason, it will fail to clean up;
                     // we'll ignore that and accept that cleanup didn't fully succeed.
                     // since we are using UUIDs for path names, this won't be an issue for
                     // snapshotting indices of the same name
-                    logger.debug(() -> new ParameterizedMessage("[{}] index [{}] no longer part of any snapshots in the repository, " +
+                    logger.error(() -> new ParameterizedMessage("[{}] index [{}] no longer part of any snapshots in the repository, " +
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: revert this change?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 10]
```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
                         "but failed to clean up its index folder due to the directory not being empty.", metadata.name(), indexId), dnee);
-                } catch (IOException ioe) {
+                } catch (Exception e) {
                     // a different IOException occurred while trying to delete - will just log the issue for now
-                    logger.debug(() -> new ParameterizedMessage("[{}] index [{}] no longer part of any snapshots in the repository, " +
-                        "but failed to clean up its index folder.", metadata.name(), indexId), ioe);
+                    logger.error(() -> new ParameterizedMessage("[{}] index [{}] no longer part of any snapshots in the repository, " +
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: revert this change?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 11]
```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
                 } catch (DirectoryNotEmptyException dnee) {
                     // if the directory isn't empty for some reason, it will fail to clean up;
                     // we'll ignore that and accept that cleanup didn't fully succeed.
                     // since we are using UUIDs for path names, this won't be an issue for
                     // snapshotting indices of the same name
-                    logger.debug(() -> new ParameterizedMessage("[{}] index [{}] no longer part of any snapshots in the repository, " +
+                    logger.error(() -> new ParameterizedMessage("[{}] index [{}] no longer part of any snapshots in the repository, " +
                         "but failed to clean up its index folder due to the directory not being empty.", metadata.name(), indexId), dnee);
-                } catch (IOException ioe) {
+                } catch (Exception e) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: revert this change?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 12]
```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
+        indicesToCleanUp.removeAll(updatedRepositoryData.getIndices().values());
+        final BlobContainer indicesBlobContainer = blobStore().blobContainer(basePath().add("indices"));
+        if (indicesToCleanUp.isEmpty()) {
+            // We're done, no indices to clean up
+            listener.onResponse(null);
+            return;
+        }
+        final AtomicInteger outstanding = new AtomicInteger(indicesToCleanUp.size());
+        for (final IndexId indexId : indicesToCleanUp) {
+            threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(() -> {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: ActionRunnable and GroupedListner

**Review by ywelsch — CHANGES_REQUESTED** [review]
In case where a snapshot deletion is running, it takes over the snapshot threadpool on the master, blocking "create snapshot" requests, which are also dispatched in TransportCreateSnapshotAction to the snapshot threadpool. I think we should have these requests come in on the generic threadpool and then go to the snapshot threadpool after actually checking on the cluster state update thread whether there is any ongoing deletes.

Also, while looking at `SnapshotsService.deleteSnapshot` I've noticed that we have callers of that method that just swallow the returned exception instead of calling the listener, grep e.g. for `deleteSnapshot(snapshot.getRepository(), snapshot.getSnapshotId().getName(), listener, true);`. Also worth fixing in a separate PR.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 8 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: @ywelsch yea I wanted to wait on #40250 for that, because we may want to push the specifics of that down to the blob container and handle the parallelization dependent on whether or not bulk deletes are supported.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 11 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: This was a conscious choice, I'd actually rather log all exceptions here directly + it simplifies things since we don't need to use the ActionRunnable here if we catch everything. I adjusted the logic a little though and invoked the grouped listener in a `finally` to make things a little clearer (and avoid the potential odd block from a future assertion trippin somewhere).

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 12 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Used `GroupedListner` but see below, I think we can do without `ActionRunnable` and keep things a little simpler.

File: server/src/main/java/org/elasticsearch/plugins/RepositoryPlugin.java [hunk 13]
```diff
@@ -39,7 +40,8 @@
      * The key of the returned {@link Map} is the type name of the repository and
      * the value is a factory to construct the {@link Repository} interface.
      */
-    default Map<String, Repository.Factory> getRepositories(Environment env, NamedXContentRegistry namedXContentRegistry) {
+    default Map<String, Repository.Factory> getRepositories(Environment env, NamedXContentRegistry namedXContentRegistry,
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/plugins/RepositoryPlugin.java]: done :)

**Comment by original-brownbear** [comment]
@ywelsch all points addressed, left ~2 questions though.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 5 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: You mean by making the methods consume a `Listener` each? That's gonna blow up the changes quite a bit I think since they are used in multiple places?

**Review by andrershov — CHANGES_REQUESTED** [review]
Seems that we need to switch TransportRestoreSnapshotAction to generic thread pool as well and then, once cluster state is checked, switch it to "snapshot" thread pool. @ywelsh do you agree?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 5 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Ah I see what you two are talking about now :D Sorry when I initially answered this I thought the criticism was about this sequence of calls:

```
            try {
                repositoryData = getRepositoryData();
                updatedRepositoryData = repositoryData.removeSnapshot(snapshotId);
                writeIndexGen(updatedRepositoryData, repositoryStateId);
            } catch (Exception ex) {
                listener.onFailure(new RepositoryException(metadata.name(), "failed to delete snapshot [" + snapshotId + "]", ex));
                return;
            }
            deleteSnapshotBlobs(snapshot, snapshotId, repositoryData, updatedRepositoryData, listener);
```

but yea, what @andrershov points out there makes more sense :)
Let me try refactoring this a bit :) sec

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 14]
```diff
@@ -236,10 +242,11 @@
      *  @param metadata       The metadata for this repository including name and settings
      * @param settings Settings for the node this repository object is created on
      */
-    protected BlobStoreRepository(RepositoryMetaData metadata, Settings settings,
-                                  NamedXContentRegistry namedXContentRegistry) {
+    protected BlobStoreRepository(RepositoryMetaData metadata, Settings settings, NamedXContentRegistry namedXContentRegistry,
+                                  ThreadPool threadPool) {
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Done :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 15]
```diff
@@ -407,66 +418,108 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
+                updatedRepositoryData = repositoryData.removeSnapshot(snapshotId);
+                writeIndexGen(updatedRepositoryData, repositoryStateId);
+            } catch (Exception ex) {
+                listener.onFailure(new RepositoryException(metadata.name(), "failed to delete snapshot [" + snapshotId + "]", ex));
+                return;
+            }
+            final SnapshotInfo finalSnapshotInfo = snapshot;
+            final Collection<IndexId> unreferencedIndices = Sets.newHashSet(repositoryData.getIndices().values());
+            unreferencedIndices.removeAll(updatedRepositoryData.getIndices().values());
+            final ActionListener<Void> afterDeleteIndices = unreferencedIndices.isEmpty()
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Turned this into a chain of callbacks now and pulled up some of the logic here to make the exact flow of what gets delete when a little clearer (I hope) and avoid multiple callsites for the lister completions in downstream.

**Comment by ywelsch** [comment]
> Seems that we need to switch TransportRestoreSnapshotAction to generic thread pool as well and then, once cluster state is checked, switch it to "snapshot" thread pool

agree

**Comment by original-brownbear** [comment]
> > Seems that we need to switch TransportRestoreSnapshotAction to generic thread pool as well and then, once cluster state is checked, switch it to "snapshot" thread pool

> agree

already done https://github.com/elastic/elasticsearch/pull/40144/files#diff-8fae4d2f6128eb9d7b26bb600da11465R53 :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 15 above)
Comment (andrershov) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Sorry, it looks even clumsier.
Is it possible to leave all the logic inside deleteIndices and deleteUnreferencedIndices? These methods can check themselves that there is no work to be done and just call the listener.
Hopefully, we can avoid ternary operator usage here.

**Review by andrershov — CHANGES_REQUESTED** [review]
I think we need to make the chain of async methods easier to read.
Regarding TransportRestoreSnapshotAction, I see that you changed action execution to a GENERIC thread pool. I don't immediately see, where restore switches back to SNAPSHOT thread pool.

**Comment by original-brownbear** [comment]
@andrershov for the threadpool thing read through `restoreSnapshot` and you'll see that it basically just does some quick "calculations" about the new metadata and then simply submits a state update task. These checks on the request and submitting that task are fine to happen on the generic pool I think.

As far as the listener, method chaining problem is concerned, take a look at the version as it is now. Thanks to #40322 I was able to massively simplify the logic for some of the delete loops and it should look more agreeable now I hope :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 16]
```diff
@@ -445,42 +445,38 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
...
+                Optional.ofNullable(finalSnapshotInfo)
+                    .map(info -> info.indices().stream().map(repositoryData::resolveIndexId).collect(Collectors.toList()))
+                    .orElse(Collections.emptyList()),
+                snapshotId,
+                ActionListener.wrap(v -> {
+                    try {
+                        blobStore().blobContainer(basePath().add("indices")).deleteBlobsIgnoringIfNotExists(
+                            unreferencedIndices.stream().map(IndexId::getId).collect(Collectors.toList()));
+                    } catch (IOException e) {
+                        // a different Exception occurred while trying to delete - will just log the issue for now
```
Comment (andrershov) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: a different Exception? Can we update the comment.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 17]
```diff
@@ -406,108 +418,99 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
+                    } catch (Exception ex) {
                         logger.warn(() ->
-                            new ParameterizedMessage("[{}] [{}] failed to read metadata for index", snapshotId, index), ex);
+                            new ParameterizedMessage("[{}] [{}] failed to read metadata for index", snapshotId, indexId.getName()), ex);
                     }
-
-                    deleteIndexMetaDataBlobIgnoringErrors(snapshot, indexId);
-
+                    deleteIndexMetaDataBlobIgnoringErrors(snapshotId, indexId);
                     if (indexMetaData != null) {
```
Comment (andrershov) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Do you think we use bulk deletes for indices? Or it's not easy because for each index we're also executing shard deletion?

**Review by andrershov — COMMENTED** [review]
I've left two comments

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 17 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: We technically could do it here already I guess, but I think I'd rather do that in another step I think. I'd rather not change the ordering (of the various deletes) here mixed with other changes, that could be a nightmare to figure out later if we have to track down an issue. 
Right now it's a little meh practically too, because it either forces us to run the loop over the indices twice because we want to delete the meta first before we touch any of the shard files.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 16 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I just removed the comment, it's pretty pointless :)

**Comment by original-brownbear** [comment]
@andrershov comments addressed :)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/bwc

**Review by andrershov — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
@andrershov thanks!
@ywelsch are you good with this one too? :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 18]
```diff
@@ -406,108 +418,98 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
-                    } catch (ElasticsearchParseException | IOException ex) {
+                    } catch (Exception ex) {
                         logger.warn(() ->
-                            new ParameterizedMessage("[{}] [{}] failed to read metadata for index", snapshotId, index), ex);
+                            new ParameterizedMessage("[{}] [{}] failed to read metadata for index", snapshotId, indexId.getName()), ex);
                     }
-
-                    deleteIndexMetaDataBlobIgnoringErrors(snapshot, indexId);
-
+                    deleteIndexMetaDataBlobIgnoringErrors(snapshotId, indexId);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: again, unrelated to the PR, but we could think in the future of moving this to a later phase, as deleting this here makes the rest unretriable.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 19]
```diff
@@ -406,108 +418,98 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
+                logger.warn(() -> new ParameterizedMessage("[{}] Unable to delete global metadata files", snapshotId), e);
+            }
+            deleteIndices(
+                Optional.ofNullable(finalSnapshotInfo)
+                    .map(info -> info.indices().stream().map(repositoryData::resolveIndexId).collect(Collectors.toList()))
+                    .orElse(Collections.emptyList()),
+                snapshotId,
+                ActionListener.wrap(v -> {
+                    try {
+                        blobStore().blobContainer(basePath().add("indices")).deleteBlobsIgnoringIfNotExists(
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: do we have tests that check that the indices folders are cleaned up for FSRepository? Might be good to add some if we don't, so that we don't inadvertently break this.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 20]
```diff
@@ -406,108 +418,98 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
+                    Arrays.asList(snapshotFormat.blobName(snapshotId.getUUID()), globalMetaDataFormat.blobName(snapshotId.getUUID())));
+            } catch (IOException e) {
+                logger.warn(() -> new ParameterizedMessage("[{}] Unable to delete global metadata files", snapshotId), e);
+            }
+            deleteIndices(
+                Optional.ofNullable(finalSnapshotInfo)
+                    .map(info -> info.indices().stream().map(repositoryData::resolveIndexId).collect(Collectors.toList()))
+                    .orElse(Collections.emptyList()),
+                snapshotId,
+                ActionListener.wrap(v -> {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: is this not equivalent to the simpler `ActionListener.map(listener, v -> { ... })`?

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left some smaller questions, looking good otherwise.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 19 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: We have a few that check this, but I'll open a PR that has more comprehensive tests for the state of the repository from https://github.com/elastic/elasticsearch/compare/master...original-brownbear:resilient-deletes-test?expand=1 (50%ish done) very shortly :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 21]
```diff
@@ -406,108 +418,98 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
...
+                return;
+            }
+            final SnapshotInfo finalSnapshotInfo = snapshot;
+            final Collection<IndexId> unreferencedIndices = Sets.newHashSet(repositoryData.getIndices().values());
+            unreferencedIndices.removeAll(updatedRepositoryData.getIndices().values());
+            try {
+                blobContainer().deleteBlobsIgnoringIfNotExists(
+                    Arrays.asList(snapshotFormat.blobName(snapshotId.getUUID()), globalMetaDataFormat.blobName(snapshotId.getUUID())));
+            } catch (IOException e) {
+                logger.warn(() -> new ParameterizedMessage("[{}] Unable to delete global metadata files", snapshotId), e);
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Sort of, we have https://github.com/elastic/elasticsearch/pull/40144/files/36b2e3888b18745edf4bb4d6e4af4c5b3165f174#diff-83e9ae765eb2a80bbbedd251b686cc10R440 writing the latest global meta-data before this happens, so we don't have a chain of references from the root to these blobs anymore regardless. But I agree, this should be retried (but as I liked in my update, that's incoming shortly too :))

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 18 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Yea, that's incoming in https://github.com/elastic/elasticsearch/compare/master...original-brownbear:delete-lock-via-cs shortly :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 12 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Right, this spot fortunately went away on its own now due to having bulk deletes :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 22]
```diff
@@ -161,6 +166,10 @@
 
     protected final RepositoryMetaData metadata;
 
+    protected final NamedXContentRegistry namedXContentRegistry;
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: my bad ... already killed it in a previous PR but it snuck back in here from a merge mistake :)

**Comment by original-brownbear** [comment]
@ywelsch all points addressed now I think :)

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks @andrershov and @ywelsch !
